### PR TITLE
Add TabWebContentsObserver & initialize WebContents

### PR DESCRIPTION
**Pull request by mshin-wolvic**
This solves the issue of gray rectagle not disappearing in the content area.
- Add TabWebContentsObserver to switch to the content view
- Initialize/show WebContents
- Fix a minor build error/crash

### CODE REVIEW

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java [hunk 1]
```diff
@@ -46,13 +46,6 @@ public void surfaceChanged(@NonNull Surface surface, int width, int height) {
 
             // Dispatch onSurfaceChanged
             mRenderView.surfaceChanged(surface, width, height);
-
-            postDelayed((Runnable) () -> {
-                @Nullable WSession.ContentDelegate delegate = mSession.getContentDelegate();
-                if (delegate != null) {
-                    delegate.onFirstComposite(mSession);
-                }
-            }, 100);
```
Comment (svillar) [inline app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java:55]: Why was this removed?

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java [hunk 2]
```diff
@@ -138,16 +147,14 @@ public void exitFullScreen() {
     @Override
     public WDisplay acquireDisplay() {
         Log.e("WolvicLifecycle", "acquire display called");
-        WebContents webContents = mRuntime.getContentShellController().createWebContents();
-        BrowserDisplay display = mRuntime.createBrowserDisplay(webContents);
-        WolvicContentRenderView renderView = mRuntime.getRenderView();
-        mRuntime.getContentShellController().getWindowAndroid().setAnimationPlaceholderView(renderView);
-        mDisplay = new DisplayImpl(display, renderView, this);
```
Comment (svillar) [inline app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java:145]: This is interesting, why this code has moved to `createBrowserDisplay()`

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java [hunk 3]
```diff
@@ -50,6 +51,14 @@ private void init() {
         mPanZoomCrontroller = new PanZoomCrontrollerImpl(this);
     }
 
+    private void registerCallbacks() {
+        mTabWebContentsObserver = new TabWebContentsObserver(mRuntime.GetCurrentWebContents(), this);
+    }
+
+    private void unRegisterCallbacks() {
+        mTabWebContentsObserver = null;
+    }
```
Comment (svillar) [inline app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java:60]: Since these are used only once, maybe we could call the code directly instead of making them private methods

File: app/src/main/res/drawable/ic_generic_file.xml [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+   - License, v. 2.0. If a copy of the MPL was not distributed with this
+   - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
+
+<vector android:height="24dp" android:viewportHeight="40"
+    android:viewportWidth="40" android:width="24dp" xmlns:android="http://schemas.android.com/apk/res/android">
+    <path android:fillColor="#FFFFFF" android:pathData="M6.5,37.5l0,-35l18.293,0l8.707,8.707l0,26.293z"/>
+    <path android:fillColor="#788B9C" android:pathData="M24.586,3L33,11.414V37H7V3H24.586M25,2H6v36h28V11L25,2L25,2z"/>
+    <path android:fillColor="#FFFFFF" android:pathData="M24.5,11.5l0,-9l0.293,0l8.707,8.707l0,0.293z"/>
+    <path android:fillColor="#788B9C" android:pathData="M25,3.414L32.586,11H25V3.414M25,2h-1v10h10v-1L25,2L25,2z"/>
+</vector>
```
Comment (svillar) [inline app/src/main/res/drawable/ic_generic_file.xml:11]: I am not sure about including this file in the location you're adding it to. Basically this file is provided by Gecko when using the Gecko backend, so if we want to add it for the Chromium backend we should do it in another location specific to Chromium

**Review by svillar — CHANGES_REQUESTED** [review]
Thanks for the patch! Looking good, I have a few questions and a request.

**Comment by mshin-wolvic** [comment]
>

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java (hunk 1 above)
Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java:55]: FYI, this branch is based on `experimental/content_api_backend`(https://github.com/Igalia/wolvic/tree/experimental/content_api_backend) branch worked by Zarhar. So, I think he could give us more of the context of this code, but I feel that this code is unnecessary since `mAfterFirstPaint` in WindowWidget.onFirstComposite that is |delegate| will still be `false` after 100 ms,  because this flag needs more time to change to `true`  for the first paint of web page.

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java (hunk 3 above)
Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java:60]: I think we will have more callbacks in `registerCallbacks` later, like mFullScreenCallback, mTabCallback.

File: app/src/main/res/drawable/ic_generic_file.xml (hunk 4 above)
Comment (mshin-wolvic) [reply app/src/main/res/drawable/ic_generic_file.xml:11]: I've added this file quickly not to have any problems compiling/launching, but it would be better to process it separately from this CL. I will remove this file.

**Comment by mshin-wolvic** [comment]
Thanks for the comments.  :-)

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java (hunk 1 above)
Comment (zakharvoit) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java:55]: This bit was originally written from the weblayer implementation so it must be outdated. Explanation above makes sense.

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java (hunk 2 above)
Comment (zakharvoit) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java:145]: I think it makes sense to not have this in SessionImpl. I would argue the best place to put this is either DisplayImpl or ContentShellController. Howevert, this seems fine for the time being while we are working on the experimental branch. I propose having a separate refactoring before starting to merge this to the main branch to define clear responsibilities for the following classes:
RuntimeImpl, SessionImpl, DisplayImpl, BrowserDisplay and ContentShellController.

right now we're focusing on single tab implementation but when we'll have potentially multiple instances of Tab/WebContents alive it will become more important.

Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java:145]: I agree with @zakharvoit. I believe this is a temporary place, and if we support the tab feature, the place may change to another place as Zakhar said.

File: app/src/main/res/drawable/ic_generic_file.xml (hunk 4 above)
Comment (duswnchl) [reply app/src/main/res/drawable/ic_generic_file.xml]: I think that we can add it under src/chromium/res once  'backend' flavor is added.

**Review by svillar — DISMISSED** [review]
Mind rebasing it on top of the current main?

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java (hunk 1 above)
Comment (duswnchl) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java]: I think that we can remove reference of mSession in DisplayImpl as well, so that  the circular dependency between DisplayImpl and Session Impl can be removed.

**Comment by mshin-wolvic** [comment]
Rebased!

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java (hunk 1 above)
Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java:55]: @duswnchl could you clean up SessionImpl's dependency when you work on the Tab support? Considering the comment below, I thought that this code would be used later when supporting the Tab. If not, we could remove it.
`
       // Make sure we update the current active tab. We have an empty SessionImpl.setActive implementation
        // because in weblayer the browser window is selected from a pool. We can use the hint that
        // the session want's to be rendered into this surface to set it active too.
`

Comment (duswnchl) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java:55]: @mshin-wolvic Aha, Sure. I'll consider it on Tab support. Thanks!

**Review by duswnchl — CHANGES_REQUESTED** [review]
Almost LGTM with some nits.

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java [hunk 5]
```diff
@@ -70,6 +70,6 @@ public WResult<Bitmap> capturePixels() {
     @Override
     public WResult<Bitmap> capturePixelsWithAspectPreservingSize(int width) {
         // TODO: Implement
-        return null;
+        return new ResultImpl<>();
```
[the remark this answers is not in the document]
Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java:73]: As far as I remember, the `null` value would lead to a crash.

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsObserver.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Override
+    public void didFirstVisuallyNonEmptyPaint() {
+        @Nullable WSession.ContentDelegate delegate = mSession.getContentDelegate();
+        if (delegate != null) {
+            delegate.onFirstContentfulPaint(mSession);
+        }
+    }
+
+}
```
[the remark this answers is not in the document]
Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsObserver.java]: Done

**Review by zakharvoit — APPROVED** [review]
LGTM

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsObserver.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import com.igalia.wolvic.browser.api.WSession;
+
+import org.chromium.content_public.browser.LifecycleState;
+import org.chromium.content_public.browser.NavigationHandle;
+import org.chromium.content_public.browser.WebContents;
+import org.chromium.content_public.browser.WebContentsObserver;
+import org.chromium.url.GURL;
+
+public class TabWebContentsObserver extends WebContentsObserver {
+    private @NonNull WebContents mWebContents;
```
Comment (jdapena) [inline app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsObserver.java:15]: I don't think you need this extra reference. mWebContents from WebContentsObserver is already set in constructor, and it is protected.

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsObserver.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        @Nullable WSession.ProgressDelegate delegate = mSession.getProgressDelegate();
+        if (delegate != null) {
+            delegate.onPageStart(mSession, url.toString());
+        }
+        dispatchCanGoBackOrForward();
+    }
+
+    @Override
+    public void didRedirectNavigation(NavigationHandle navigationHandle) {
+        dispatchCanGoBackOrForward();
```
Comment (jdapena) [inline app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsObserver.java:35]: Don't you need to update the progress delegate with the new URL? Maybe with a new onPageStart session?

Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsObserver.java:35]: There is no suitable API like onPageRedirect in  WSession.ProgressDelegate and |url| is not used in all listeners of onPageStart. And I'm worried we will call multiple onPageStart since TelemetryService could be affected.

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsObserver.java (hunk 7 above)
Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsObserver.java]: Done!

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsObserver.java (hunk 8 above)
Comment (jdapena) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsObserver.java:33]: Fair enough

**Review by jdapena — APPROVED** [review]
